data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -226,10 +266,14 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
 
     // if there are fewer deletes than a reasonable number to keep in memory, use a set
     if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
-      return Deletes.filter(records, this::pos, Deletes.toPositionIndex(filePath, deletes));
+      return hasMetadataColumnIsDeleted ?
+          Deletes.marker(records, this::pos, Deletes.toPositionIndex(filePath, deletes), this::markRowDeleted) :
```

Comment:
I feel we should abstract these later. 
```
if (needWrapDeleteColumn) {
    Deletes.map(records, this::pos, Deletes.toPositionoIndex(filePath, deletes), this::markRowDeleted);
} else {
    Deletes.filter(..)
}
```

What about adding a class DeleteStream that supports `filter` and `map`.

Reply:
The `DeleteStream` contains the filter logics in this `DeleteFilter` also contains transforming logics.

Reply:
We can definitely refactor here. I've got some ideas in my mind. Trying to understand your point here. Do you mean a class `DeleteStream` combine both functionality from `PositionSetDeleteFilter` and `PositionSetDeleteMarker`? I will test these ideas anyway. Thanks for the suggestion.

Reply:
Yes, like the Java steam API.